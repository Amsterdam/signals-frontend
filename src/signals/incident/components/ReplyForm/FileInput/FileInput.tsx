// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2021 - 2022 Vereniging van Nederlandse Gemeenten, Gemeente Amsterdam
import { useCallback, useState } from 'react'

import type { Control } from 'react-hook-form'
import { useController } from 'react-hook-form'

import FileInputComponent from 'components/FileInput'
import fileSize from 'signals/incident/services/file-size'

export const MIN = 30 * 2 ** 10 // 30 KiB
const MAX = 20 * 2 ** 20 // 20 MiB
const MAX_NUMBER_OF_FILES = 3
export const ALLOWED_FILE_TYPES = [
  'image/jpeg',
  'image/jpg',
  'image/png',
  'image/gif',
]
const ALLOWED_EXTENSIONS = ALLOWED_FILE_TYPES.map((type) => type.split('/')[1])

type FileInputProps = {
  errorMessage?: string
  shortLabel: string
  id: string
  label: string
  trigger: (id: string) => void
  control: Control<Record<string, unknown>>
}

const FileInput = ({
  errorMessage,
  shortLabel,
  id,
  label,
  trigger,
  control,
}: FileInputProps) => {
  const [files, setFiles] = useState<File[]>([])
  const controller = useController({
    control,
    name: id,
    defaultValue: [],
    rules: {
      validate: {
        fileType: (files) => {
          if (
            (files as File[]).find(
              (file) => !ALLOWED_FILE_TYPES.includes(file.type)
            )
          ) {
            return `Dit bestandstype wordt niet ondersteund. Toegestaan zijn: ${ALLOWED_EXTENSIONS.join(
              ', '
            )}`
          }
        },
        numberOfFiles: (files) => {
          if ((files as File[]).length > MAX_NUMBER_OF_FILES)
            return `U kunt maximaal ${MAX_NUMBER_OF_FILES} bestanden uploaden`
        },
        minFileSize: (files) =>
          (files as File[]).find((file) => file.size < MIN) &&
          `Dit bestand is te klein. De minimale bestandsgrootte is ${fileSize(
            MIN
          )}.`,
        maxFileSize: (files) =>
          (files as File[]).find((file) => file.size > MAX) &&
          `Dit bestand is te groot. De maximale bestandsgrootte is ${fileSize(
            MAX
          )}.`,
      },
    },
  })

  const handleChange = useCallback(
    (files: File[]) => {
      controller.field.onChange(files)
      setFiles(files)
      trigger(id)
    },
    [controller.field, id, trigger]
  )

  return (
    <FileInputComponent
      name={id}
      label={shortLabel}
      helpText={label}
      errorMessages={errorMessage ? [errorMessage] : []}
      onChange={handleChange}
      files={files}
    />
  )
}

export default FileInput
