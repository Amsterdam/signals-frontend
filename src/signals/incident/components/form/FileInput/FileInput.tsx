// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2018 - 2023 Gemeente Amsterdam
import { useState, useCallback } from 'react'

import FileInputComponent from 'components/FileInput'

import fileSize from '../../../services/file-size'
import type { Meta, Parent } from '../types/FileInput'

export interface Props {
  handler: () => { value: File[] }
  parent: Parent
  meta: Meta
}

const FileInput = ({ handler, parent, meta }: Props) => {
  const [errors, setErrors] = useState<string[]>()
  const maxNumberOfFiles = (meta && meta.maxNumberOfFiles) || 3
  const checkMinFileSize = useCallback(
    (file) => file.size >= meta.minFileSize,
    [meta]
  )

  const checkMaxFileSize = useCallback(
    (file) => file.size < meta.maxFileSize,
    [meta]
  )

  const checkFileType = useCallback(
    (file) => meta.allowedFileTypes.includes(file.type),
    [meta]
  )

  const checkNumberOfFiles = useCallback(
    (_, index) => index < maxNumberOfFiles,
    [maxNumberOfFiles]
  )

  const getErrorMessages = useCallback(
    (files) => {
      const errorMessages = []

      if (meta.minFileSize && !files.every(checkMinFileSize)) {
        errorMessages.push(
          `Dit bestand is te klein. De minimale bestandgrootte is ${fileSize(
            meta.minFileSize
          )}.`
        )
      }

      if (meta.maxFileSize && !files.every(checkMaxFileSize)) {
        errorMessages.push(
          `Dit bestand is te groot. De maximale bestandgrootte is ${fileSize(
            meta.maxFileSize
          )}.`
        )
      }

      if (meta.allowedFileTypes && !files.every(checkFileType)) {
        errorMessages.push(
          `Dit bestandstype wordt niet ondersteund. Toegestaan zijn: ${meta.allowedFileTypes
            .map((type) => type.replace(/.*\//, ''))
            .join(', ')}.`
        )
      }

      if (!files.every(checkNumberOfFiles)) {
        errorMessages.push(
          `U kunt maximaal ${maxNumberOfFiles} bestanden uploaden.`
        )
      }

      return errorMessages
    },
    [
      meta,
      checkMinFileSize,
      checkMaxFileSize,
      checkFileType,
      checkNumberOfFiles,
      maxNumberOfFiles,
    ]
  )

  const handleChange = useCallback(
    (batchFiles) => {
      const minFileSizeFilter = meta.minFileSize ? checkMinFileSize : () => true
      const maxFileSizeFilter = meta.maxFileSize ? checkMaxFileSize : () => true
      const allowedFileTypesFilter = meta.allowedFileTypes
        ? checkFileType
        : () => true
      const maxNumberOfFilesFilter = checkNumberOfFiles

      const files = batchFiles
        .filter(minFileSizeFilter)
        .filter(maxFileSizeFilter)
        .filter(allowedFileTypesFilter)
        .filter(maxNumberOfFilesFilter)

      setErrors(getErrorMessages(batchFiles))

      parent.meta.updateIncident({
        [meta.name]: files,
        [`${meta.name}_previews`]: files.map((file: File) =>
          window.URL.createObjectURL(file)
        ),
      })
    },
    [
      checkMinFileSize,
      checkMaxFileSize,
      checkFileType,
      checkNumberOfFiles,
      getErrorMessages,
      meta,
      parent,
    ]
  )

  const files = handler().value || []

  return (
    <FileInputComponent
      allowedFileTypes={meta.allowedFileTypes}
      errorMessages={errors}
      files={files}
      maxFileSize={meta.maxFileSize}
      maxNumberOfFiles={maxNumberOfFiles}
      minFileSize={meta.minFileSize}
      name={meta.name}
      onChange={handleChange}
    />
  )
}

export default FileInput
