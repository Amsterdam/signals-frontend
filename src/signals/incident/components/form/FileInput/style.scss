.file-input {
  &__preview {
    &-image {
      border: 1px solid black;
      height: 80px;
    }
  }
}
