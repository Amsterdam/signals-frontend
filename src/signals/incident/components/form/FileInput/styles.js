import styled from 'styled-components';
import { themeColor, themeSpacing, Icon } from '@datapunt/asc-ui';
import Button from 'components/Button';

const FileInputStyle = styled.div`
  display: flex;
  height: ${themeSpacing(25)};
`;

export const FileInputPreviewBox = styled.div`
  width: ${themeSpacing(25)};
  margin-right: ${themeSpacing(3)};
`;

export const FileInputEmptyBox = styled.div`
  width: ${themeSpacing(25)};
  border: 1px dashed ${themeColor('tint', 'level4')};
  margin-right: ${themeSpacing(3)};
`;

export const FileInputError = styled.div`
  color: ${themeColor('secondary')};
  margin: ${themeSpacing(4, 0, 0)};
`;

export const FileInputUploadButton = styled.div`
  width: ${themeSpacing(25)};
  border: 1px dashed ${themeColor('tint', 'level4')};
  margin-right: ${themeSpacing(3)};

  input[type='file'] {
    opacity: 0;
    width: 0;
    height: 0;
  }

  & > label {
    position: relative;
    display: inline-block;
    cursor: pointer;
    height: 100%;
    width: 100%;
  }
`;

export const DeleteButton = styled(Button)`
  position: absolute;
  width: 40px;
  height: 40px;
  bottom: 0;
  right: 0;
  background-color: rgba(0, 0, 0, 0.5);

  svg > path {
    fill: ${themeColor('tint', 'level1')};
  }

  &:hover {
    background-color: black;
  }
`;

export const AddButton = styled.div`
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  width: ${themeSpacing(11)};
  height: ${themeSpacing(11)};
  top: ${themeSpacing(7)};
  left: ${themeSpacing(7)};
  border-radius: 50%;
  border: 1px solid ${themeColor('primary')};

  &:hover {
    background-color: ${themeColor('tint', 'level1')};
    border-width: 2px;
    padding: 2px;
  }
`;

export const AddIcon = styled(Icon)`
  svg > path {
    fill: ${themeColor('primary')};
  }
`;

export const FilePreview = styled.div`
  position: relative;
  background-size: cover;
  height: 100%;
  width: 100%;
  background-image: ${({ preview }) => `URL(${preview})`};
`;

export const FileLoading = styled.div`
  position: relative;
  height: 100%;
  top: ${themeSpacing(9)};
  left: ${themeSpacing(5)};
`;

export default FileInputStyle;
