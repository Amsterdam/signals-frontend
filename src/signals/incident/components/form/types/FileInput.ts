import type { WizardSection } from 'signals/incident/definitions/wizard'
import type { FormMeta } from 'types/reactive-form'

export enum FileTypes {
  JPEG = 'image/jpeg',
  JPG = 'image/jpg',
  PNG = 'image/png',
  GIF = 'image/gif',
}

export interface Meta extends FormMeta {
  allowedFileTypes: FileTypes[]
  label: string
  maxFileSize: number
  maxNumberOfFiles: number
  minFileSize: number
  name: string
  subtitle: string
  isVisible?: boolean
}

export interface ParentMeta {
  updateIncident: (data: any) => void
  addToSelection?: () => void
  getClassification?: () => void
  handleSubmit?: () => void
  incidentsContainer?: {
    incident: {
      id: number
    }
  }
  removeFromSelection?: () => void
  wizard?: WizardSection
}

export interface Parent {
  meta: ParentMeta
  value?: any
}
