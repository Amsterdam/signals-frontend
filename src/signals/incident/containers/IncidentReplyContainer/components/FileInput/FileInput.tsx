// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2021 - 2022 Vereniging van Nederlandse Gemeenten, Gemeente Amsterdam
import FileInputComponent from 'components/FileInput'
import type { FunctionComponent } from 'react'
import { useCallback, useState } from 'react'
import { useController } from 'react-hook-form'
import fileSize from 'signals/incident/services/file-size'
import type { FieldProps } from '../../types'

const MIN = 30 * 2 ** 10 // 30 KiB
const MAX = 20 * 2 ** 20 // 20 MiB
const MAX_NUMBER_OF_FILES = 3
const ALLOWED_FILE_TYPES = ['image/jpeg', 'image/jpg', 'image/png', 'image/gif']
const ALLOWED_EXTENSIONS = ALLOWED_FILE_TYPES.map((type) => type.split('/')[1])

const FileInput: FunctionComponent<FieldProps> = ({
  errorMessage,
  id,
  label,
  trigger,
  control,
}) => {
  const [files, setFiles] = useState<File[]>([])
  const controller = useController({
    control,
    name: id,
    defaultValue: [],
    rules: {
      validate: {
        fileType: (files) => {
          if (
            (files as File[]).find(
              (file) => !ALLOWED_FILE_TYPES.includes(file.type)
            )
          ) {
            return `Dit bestandstype wordt niet ondersteund. Toegestaan zijn: ${ALLOWED_EXTENSIONS.join(
              ', '
            )}`
          }
        },
        numberOfFiles: (files) => {
          if ((files as File[]).length > MAX_NUMBER_OF_FILES)
            return `U kunt maximaal ${MAX_NUMBER_OF_FILES} bestanden uploaden`
        },
        minFileSize: (files) =>
          (files as File[]).find((file) => file.size < MIN) &&
          `Dit bestand is te klein. De minimale bestandsgrootte is ${fileSize(
            MIN
          )}.`,
        maxFileSize: (files) =>
          (files as File[]).find((file) => file.size > MAX) &&
          `Dit bestand is te groot. De maximale bestandsgrootte is ${fileSize(
            MAX
          )}.`,
      },
    },
  })

  const handleChange = useCallback(
    (files: File[]) => {
      controller.field.onChange(files)
      setFiles(files)
      trigger(id)
    },
    [controller.field, id, trigger]
  )

  return (
    <FileInputComponent
      name={id}
      label={label}
      helpText="Voeg een foto toe om de situatie te verduidelijken."
      errorMessages={errorMessage ? [errorMessage] : []}
      onChange={handleChange}
      files={files}
    />
  )
}

export default FileInput
