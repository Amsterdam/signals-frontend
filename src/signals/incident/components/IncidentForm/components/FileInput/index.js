import React from 'react';
import PropTypes from 'prop-types';

import Header from '../Header/';

import './style.scss';

const FileInput = ({ handler, touched, hasError, getError, parent, meta, validatorsOrOpts }) => {
  const handleChange = (e) => {
    if (e.target.files && e.target.files.length) {
      const file = e.target.files[0];

      // use revokeObjectURL afterward
      const url = window.URL.createObjectURL(file);
      parent.meta.setIncident({
        image: url
      });

      const reader = new window.FileReader();
      reader.addEventListener('load', () => {
        parent.meta.setIncident({
          image_file: file
        });
      });

      reader.readAsText(file);
    }
  };

  const handleClear = (url) => {
    window.URL.revokeObjectURL(url);
    parent.meta.setIncident({
      image: '',
      image_file: {}
    });
  };

  return (
    <div className={`${meta && meta.isVisible ? 'row' : ''}`}>
      {meta && meta.isVisible ?
        <div className={`${meta.className || 'col-12'} mode_input file`}>
          <Header
            meta={meta}
            options={validatorsOrOpts}
            touched={touched}
            hasError={hasError}
            getError={getError}
          >
            {handler().value ?
              <div className="file-input__preview">
                <button
                  className="file-input__button-delete link-functional delete"
                  onClick={() => handleClear(handler().value)}
                />

                <img
                  alt="Preview uploaded foto"
                  src={handler().value}
                  className="file-input__preview-image"
                />
              </div>
            :
              <div className="invoer">
                <input
                  type="file"
                  id="formUpload"
                  onChange={handleChange}
                />
                <label htmlFor="formUpload" className="secundary-blue">{meta.submitLabel}</label>
              </div>
            }
          </Header>
        </div>
         : ''}
    </div>
  );
};

FileInput.propTypes = {
  handler: PropTypes.func,
  touched: PropTypes.bool,
  hasError: PropTypes.func,
  meta: PropTypes.object,
  parent: PropTypes.object,
  getError: PropTypes.func.isRequired,
  validatorsOrOpts: PropTypes.object
};

export default FileInput;
