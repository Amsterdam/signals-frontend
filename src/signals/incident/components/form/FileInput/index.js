import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import Header from '../Header';
import fileSize from '../../../services/file-size';

import './style.scss';

export const ERROR_TIMEOUT_INTERVAL = 8000;

const FileInput = ({
  handler, touched, hasError, getError, parent, meta, validatorsOrOpts,
}) => {
  let timeoutInstance = null;
  const maxNumberOfFiles = (meta && meta.maxNumberOfFiles) || 3;
  const handleChange = e => {
    /* istanbul ignore next */
    if (e.target.files && e.target.files.length) {
      const minFileSizeFilter = meta.minFileSize ? checkMinFileSize : () => true;
      const maxFileSizeFilter = meta.maxFileSize ? checkMaxFileSize : () => true;
      const allowedFileTypesFilter = meta.allowedFileTypes ? checkFileType : () => true;
      const maxNumberOfFilesFilter = checkNumberOfFiles;
      const existingFiles = handler().value || [];
      const existingPreviews = (parent && parent.value && parent.value[`${meta.name}_previews`]) || [];
      const batchFiles = [...e.target.files];

      existingFiles.map(file => ({ ...file, existing: true }));
      const files = [...existingFiles, ...batchFiles]
        .filter(minFileSizeFilter)
        .filter(maxFileSizeFilter)
        .filter(allowedFileTypesFilter)
        .filter(maxNumberOfFilesFilter);

      const previews = [...existingPreviews, ...batchFiles.map(() => `loading-${Math.trunc(Math.random() * 100000)}`)]
        .slice(0, files.length);
      const errors = getErrorMessages([...existingFiles, ...batchFiles]);
      parent.meta.updateIncident({
        [meta.name]: files,
        [`${meta.name}_previews`]: previews,
        [`${meta.name}_errors`]: errors,
      });

      if (errors.length) {
        if (timeoutInstance) {
          global.window.clearTimeout(timeoutInstance);
        }
        timeoutInstance = global.window.setTimeout(() => {
          parent.meta.updateIncident({
            [`${meta.name}_errors`]: null,
          });
        }, ERROR_TIMEOUT_INTERVAL);
      }

      files.forEach((file, uploadBatchIndex) => {
        if (files[uploadBatchIndex].existing) {
          return;
        }
        const reader = new window.FileReader();
        reader.addEventListener('load', () => {
          previews[uploadBatchIndex] = window.URL.createObjectURL(files[uploadBatchIndex]);
          parent.meta.updateIncident({
            [`${meta.name}_previews`]: previews,
          });

          const control = meta && meta.name && parent.controls[meta.name];
          control.updateValueAndValidity();
        });

        reader.readAsText(file);
      });
    }
  };

  const checkMinFileSize = file => file.size >= meta.minFileSize;

  const checkMaxFileSize = file => file.size < meta.maxFileSize;

  const checkFileType = file => meta.allowedFileTypes.includes(file.type);

  const checkNumberOfFiles = (file, index) => index < maxNumberOfFiles;

  const getErrorMessages = files => {
    const errors = [];

    if (meta.minFileSize && !files.every(checkMinFileSize)) {
      errors.push(`Dit bestand is te klein. De minimale bestandgrootte is ${fileSize(meta.minFileSize)}.`);
    }
    if (meta.maxFileSize && !files.every(checkMaxFileSize)) {
      errors.push(`Dit bestand is te groot. De maximale bestandgrootte is ${fileSize(meta.maxFileSize)}.`);
    }
    if (meta.allowedFileTypes && !files.every(checkFileType)) {
      errors.push(`Dit bestandstype wordt niet ondersteund. Toegestaan zijn: ${meta.allowedFileTypes.map(type => type.replace(/.*\//, '')).join(', ')}.`);
    }
    if (!files.every(checkNumberOfFiles)) {
      errors.push(`U kunt maximaal ${maxNumberOfFiles} bestanden uploaden.`);
    }

    return errors;
  };

  const removeFile = (e, preview, previews, files) => {
    e.preventDefault();

    const key = previews.indexOf(preview);
    /* istanbul ignore next */
    if (key !== -1) {
      window.URL.revokeObjectURL(preview);

      files.splice(key, 1);
      previews.splice(key, 1);

      parent.meta.updateIncident({
        [meta.name]: files,
        [`${meta.name}_previews`]: previews,
        [`${meta.name}_errors`]: null,
      });
    }

    const control = meta && meta.name && parent.controls[meta.name];
    control.updateValueAndValidity();
  };

  const previews = (parent && parent.value && parent.value[`${meta && meta.name}_previews`]) || [];
  const errors = (parent && parent.value && parent.value[`${meta && meta.name}_errors`]) || null;
  const numberOfEmtpy = maxNumberOfFiles - previews.length - 1;
  const empty = numberOfEmtpy < 0 ? [] : Array.from(Array(numberOfEmtpy).keys());

  if (!meta?.isVisible) return null;

  return (
    <Fragment>
      <Header
        meta={meta}
        options={validatorsOrOpts}
        touched={touched}
        hasError={hasError}
        getError={getError}
      >
        <div className="file-input">
          {previews.length > 0 && previews.map(preview => (
            <div key={preview} className={`file-input__preview ${preview.includes('loading') ? 'file-input__preview--loading' : ''}`}>
              {preview.includes('loading')
                ? <div className="progress-indicator progress-red"></div>
                : (
                  <div style={{ backgroundImage: `URL(${preview})` }} className="file-input__preview-image">
                    <button aria-label="Verwijder deze foto" type="button" className="file-input__preview-button-delete" onClick={e => removeFile(e, preview, previews, handler().value)} />
                  </div>
                )}
            </div>
          ))}

          {previews.length < maxNumberOfFiles && (
            <div className="file-input__button">
              <label htmlFor="formUpload" className="file-input__button-label">
                <div className="file-input__button-label-icon" />
              </label>
              <input
                type="file"
                id="formUpload"
                accept={meta.allowedFileTypes}
                onChange={handleChange}
                multiple
              />
            </div>
          )}

          {empty.map(item => (<div key={item} className="file-input__empty">&nbsp;</div>))}
        </div>
      </Header>
      {errors?.length > 0 && errors.map(error => <div key={error} className="file-input__error">{error}</div>)}
    </Fragment>
  );
};

FileInput.propTypes = {
  handler: PropTypes.func,
  touched: PropTypes.bool,
  hasError: PropTypes.func,
  meta: PropTypes.object,
  parent: PropTypes.object,
  getError: PropTypes.func.isRequired,
  validatorsOrOpts: PropTypes.object,
};

export default FileInput;
