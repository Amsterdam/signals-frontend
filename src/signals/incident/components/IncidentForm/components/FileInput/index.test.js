import React from 'react';
import { shallow } from 'enzyme';

import FileInput from './index';

describe('Form component <FileInput />', () => {
  const metaFields = {
    name: 'input-field-name',
    submitLabel: 'upload file'
  };
  let wrapper;
  let handler;
  let touched;
  let getError;
  let hasError;
  let parent;

  beforeEach(() => {
    handler = jest.fn();
    touched = false;
    getError = jest.fn();
    hasError = jest.fn();
    parent = {
      meta: {
        setIncident: jest.fn()
      }
    };

    wrapper = shallow(<FileInput
      handler={handler}
      parent={parent}
      touched={touched}
      hasError={hasError}
      getError={getError}
    />);
  });

  describe('rendering', () => {
    it('should render upload field correctly', () => {
      handler.mockImplementation(() => ({ value: undefined }));

      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true
        }
      });

      expect(handler).toHaveBeenCalledWith();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render upload field with selected file correctly', () => {
      handler.mockImplementation(() => ({ value: 'blob:http://host/c00d2e14-ae1c-4bb3-b67c-86ea93130b1c' }));

      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true
        }
      });

      expect(handler).toHaveBeenCalledWith();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render no upload field when not visible', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: false
        }
      });

      expect(handler).not.toHaveBeenCalled();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('events', () => {
    const fileContents = 'file contents';
    const file = new Blob([fileContents], { type: 'text/plain' });
    let readAsText;
    let addEventListener;

    beforeEach(() => {
      readAsText = jest.fn();
      addEventListener = jest.fn((_, evtHandler) => { evtHandler(); });
      window.FileReader = jest.fn(() => ({
        addEventListener,
        readAsText,
        result: fileContents
      }));

      window.URL = {
        createObjectURL: jest.fn(),
        revokeObjectURL: jest.fn()
      };
    });

    afterEach(() => {
      delete window.URL;
    });

    it('uploads a file when and sets incident when file changes', () => {
      handler.mockImplementation(() => ({ value: undefined }));

      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true
        }
      });

      wrapper.find('input').simulate('change', { target: { files: [] } });
      expect(FileReader).not.toHaveBeenCalled();

      wrapper.find('input').simulate('change', { target: { files: [file] } });
      expect(FileReader).toHaveBeenCalled();
      expect(addEventListener).toHaveBeenCalledWith('load', expect.any(Function));
      expect(readAsText).toHaveBeenCalledWith(file);
    });

    it('resets upload when clear button was clicked', () => {
      handler.mockImplementation(() => ({ value: 'blob:http://host/c00d2e14-ae1c-4bb3-b67c-86ea93130b1c' }));

      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true
        }
      });

      wrapper.find('button').simulate('click');

      expect(window.URL.revokeObjectURL).toHaveBeenCalledWith('blob:http://host/c00d2e14-ae1c-4bb3-b67c-86ea93130b1c');
      expect(parent.meta.setIncident).toHaveBeenCalledWith({
        image: '',
        image_file: {}
      });
    });
  });
});
