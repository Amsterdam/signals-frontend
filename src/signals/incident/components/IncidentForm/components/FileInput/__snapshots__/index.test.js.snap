// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component <FileInput /> rendering should render no upload field when not visible 1`] = `
<div
  className=""
/>
`;

exports[`Form component <FileInput /> rendering should render upload field correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input file"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "name": "input-field-name",
          "submitLabel": "upload file",
        }
      }
      touched={false}
    >
      <div
        className="invoer"
      >
        <input
          id="formUpload"
          onChange={[Function]}
          type="file"
        />
        <label
          className="secundary-blue"
          htmlFor="formUpload"
        >
          upload file
        </label>
      </div>
    </Header>
  </div>
</div>
`;

exports[`Form component <FileInput /> rendering should render upload field with selected file correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input file"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "name": "input-field-name",
          "submitLabel": "upload file",
        }
      }
      touched={false}
    >
      <div
        className="file-input__preview"
      >
        <button
          className="file-input__button-delete link-functional delete"
          onClick={[Function]}
        />
        <img
          alt="Preview uploaded foto"
          className="file-input__preview-image"
          src="blob:http://host/c00d2e14-ae1c-4bb3-b67c-86ea93130b1c"
        />
      </div>
    </Header>
  </div>
</div>
`;
